damnit/api.py

```diff
@@ -171,6 +173,110 @@ def summary(self):
                 return blob2complex(value)
             return value
 
+    def preview_data(self, *, data_fallback=True, deserialize_plotly=True):
+        """Get the preview data for the variable
+
+        May return a 1D or 2D data array, a 3D RGB(A) arrray, a plotly figure
+        object, a str of plotly JSON (with deserialize_plotly=False) or None
+        if no preview is available.
+
+        If no preview was specified in the context file, but the returned data
+        meets the conditions for a preview, this can be returned instead.
+        Pass `data_fallback=False` to prevent this.
+        """
+        with h5py.File(self._h5_path) as f:
+            xarray_group = dset = None
+            if (obj := f.get(f".preview/{self.name}")) is not None:
+                # Explicit preview
+                type_hint = self._type_hint(obj)
+                if isinstance(obj, h5py.Group):
+                    xarray_group = obj.name
+                else:
+                    dset = obj
+            elif data_fallback:
+                # Implicit: use data as preview if suitable
+                grp = f[self.name]
+                type_hint = self._type_hint(grp)
+                if type_hint is DataType.DataArray:
+                    xarray_group = self.name
+                elif type_hint is DataType.Dataset:
+                    return None
+                else:
+                    dset = grp['data']
+            else:
+                return None
+
+            if xarray_group is not None:
+                for obj in f[xarray_group].values():
+                    if isinstance(obj, h5py.Dataset) and (
+                        obj.ndim > 3 or (obj.ndim == 3 and obj.shape[-1] not in (3, 4))
+                    ):
+                        return None  # Too many dims: bail out before loading
+
+                import xarray as xr
+                arr = xr.load_dataarray(
+                    self._h5_path, group=xarray_group, engine="h5netcdf"
+                )
+                if arr.ndim != 0 and np.issubdtype(arr.dtype, np.number):
+                    return arr
+
+            elif np.issubdtype(dset.dtype, np.number) and (
+                    dset.ndim in (1, 2) or (dset.ndim == 3 and dset.shape[-1] in (3, 4))
+            ):
+                value = dset[()]
+
+                if type_hint is DataType.PlotlyFigure:
+                    import plotly.io as pio
+                    b = value.tobytes()
+                    return pio.from_json(b) if deserialize_plotly else b.decode()
+                else:
+                    return value
+
+        return None
+
+    def preview(self):
+        """Show the preview data
+
+        This is intended for use in Jupyter notebooks
+        """
+        if (obj := self.preview_data()) is None:
+            return
```

Comment:
Could we return a string like "Preview could not be generated"? That would be a bit clearer than the function seeming to do nothing when called.

Reply:
I agree it makes sense to have something visible in that case; what do you think about throwing an error rather than returning a string?

Reply:
Yep that seems fine.